cephadm: Indicate orch is paused on executing operations

Kushal-deb opened the pull request:
cephadm: Indicate orch is paused on executing operations. 

This fix introduces a decorator, `check_orchestrator_paused`, to ensure that commands executed when the orchestrator is paused provide a clear warning to the user.

Fixes: https://tracker.ceph.com/issues/69219

Test logs:

```
[ceph: root@ceph-node-0 ~]# ceph orch status
Backend: cephadm
Available: Yes
Paused: Yes

[ceph: root@ceph-node-0 ~]# ceph orch rm crash
WARNING: The following command will be deferred and executed once the orchestrator is resumed.
Removed service crash

[ceph: root@ceph-node-0 ~]# ceph orch ls
NAME                       PORTS  RUNNING  REFRESHED   AGE  PLACEMENT  
crash                                 3/3  <deleting>  9m   *          
mgr                                   2/2  4m ago      9m   count:2    
mon                                   3/5  4m ago      9m   count:5    
osd.all-available-devices               3  4m ago      8m   *    
```      

```
[ceph: root@ceph-node-0 ~]# ceph orch status
Backend: cephadm
Available: Yes
Paused: No

[ceph: root@ceph-node-0 ~]# ceph orch ls
NAME                       PORTS  RUNNING  REFRESHED  AGE  PLACEMENT  
mgr                                   2/2  1s ago     9m   count:2    
mon                                   3/5  1s ago     9m   count:5    
osd.all-available-devices               3  1s ago     8m   *      
    
[ceph: root@ceph-node-0 ~]# ceph orch apply crash
Scheduled crash update...
```


## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

phlogistonjohn commented on src/pybind/mgr/orchestrator/module.py:
```diff
@@ -2381,20 +2417,23 @@ def _upgrade_start(self,
         return HandleCommandResult(stdout=completion.result_str())
 
     @_cli_write_command('orch upgrade stop')
+    @check_orchestrator_paused(get_status_func=lambda self, **kwargs: self._status(**kwargs))
```
I do object to this repeated lambda stuff. if you have to repeat this on nearly every invocation of your dectorator, build this into the decorator do not "spam" the codebase with it.

Kushal-deb replied:
Removed it in the latest commit.

phlogistonjohn reviewed (changes requested):
I won't block this getting merged, but I will say that it's a bit ugly due to the "overuse" of HandleCommandResult  (and the subsequent string manipulations).  If cephadm were to adopt an approach of using more typical python workflows with (specific) exception types (an approach encouraged by the object formatter module) this could be much shorter and less ugly.

Kushal-deb commented:
> I won't block this getting merged, but I will say that it's a bit ugly due to the "overuse" of HandleCommandResult (and the subsequent string manipulations). If cephadm were to adopt an approach of using more typical python workflows with (specific) exception types (an approach encouraged by the object formatter module) this could be much shorter and less ugly.

Hi @phlogistonjohn,

Thank you for your earlier feedback! The current commit addresses your concern about the overuse of HandleCommandResult and the clutter caused by the repeated invocation of lambdas. 

I would, however, like to seek some clarification regarding the behavior when the orchestrator is paused. For example, in its current state, if we attempt to remove a service like crash (e.g., ceph orch rm crash), the system places it in a "deleting" state, and the removal is deferred until the orchestrator is resumed. In contrast, other commands like ceph orch daemon rm osd.6 --force don't exhibit this behavior—they immediately remove the daemon without deferring. Similarly, the ceph orch tuned-profile apply command also applies the changes directly, even when the orchestrator is paused. I've included relevant output snippets below to provide further context.

Given this inconsistency, would it be more appropriate to simplify the behavior by raising an exception when the orchestrator is paused and displaying a message like:
"Orchestrator is paused. Please try again after resuming the orchestrator."

Alternatively, should we aim to ensure consistency by deferring all such operations during the paused state? I’d appreciate your feedback on this approach, and I’m happy to iterate further based on your suggestions.

**Removing crash**
```
[ceph: root@ceph-node-0 ~]# ceph orch status
Backend: cephadm
Available: Yes
Paused: Yes

[ceph: root@ceph-node-0 ~]# ceph orch rm crash
Error EAGAIN: WARNING: The orchestrator is paused. The command is deferred for execution until orchestrator is resumed. 
Removed service crash

[ceph: root@ceph-node-0 ~]# ceph orch ls
NAME                       PORTS  RUNNING  REFRESHED   AGE  PLACEMENT  
crash                                 4/4  <deleting>  51m  *          
```

**Removing an OSD daemon:**

```
[ceph: root@ceph-node-0 ~]# ceph orch ps 
NAME                    HOST         PORTS        STATUS         REFRESHED  AGE  MEM USE  MEM LIM  VERSION                IMAGE ID      CONTAINER ID  
osd.0                   ceph-node-0               running (26h)    64s ago   2d    60.7M    4096M  19.3.0-6008-g60ae04d8  a8b4c1808756  0b9d1256d028  
osd.1                   ceph-node-1               running (26h)    64s ago   2d    63.4M    4096M  19.3.0-6008-g60ae04d8  a8b4c1808756  ba2a1dffefa7  
osd.2                   ceph-node-2               running (26h)    64s ago   2d    63.6M    1496M  19.3.0-6008-g60ae04d8  a8b4c1808756  449a16053d5f  
osd.3                   ceph-node-0               running (26h)    64s ago   2d    63.2M    4096M  19.3.0-6008-g60ae04d8  a8b4c1808756  50aca0b19f7c  
osd.4                   ceph-node-1               running (26h)    64s ago   2d    60.3M    4096M  19.3.0-6008-g60ae04d8  a8b4c1808756  1560f80ba29a  
osd.5                   ceph-node-2               running (26h)    64s ago   2d    60.5M    1496M  19.3.0-6008-g60ae04d8  a8b4c1808756  6354be288c00  
osd.6                   ceph-node-3               running (23h)    64s ago  23h    63.7M    2992M  19.3.0-6008-g60ae04d8  a8b4c1808756  020aa018ae78  

[ceph: root@ceph-node-0 ~]# ceph orch daemon rm osd.6 --force
Error EAGAIN: WARNING: The orchestrator is paused. The command is deferred for execution until orchestrator is resumed. 
Removed osd.6 from host 'ceph-node-3'

[ceph: root@ceph-node-0 ~]# ceph orch ps
NAME                    HOST         PORTS        STATUS         REFRESHED  AGE  MEM USE  MEM LIM  VERSION                IMAGE ID      CONTAINER ID  
osd.0                   ceph-node-0               running (26h)    94s ago   2d    60.7M    4096M  19.3.0-6008-g60ae04d8  a8b4c1808756  0b9d1256d028  
osd.1                   ceph-node-1               running (26h)    94s ago   2d    63.4M    4096M  19.3.0-6008-g60ae04d8  a8b4c1808756  ba2a1dffefa7  
osd.2                   ceph-node-2               running (26h)    94s ago   2d    63.6M    1496M  19.3.0-6008-g60ae04d8  a8b4c1808756  449a16053d5f  
osd.3                   ceph-node-0               running (26h)    94s ago   2d    63.2M    4096M  19.3.0-6008-g60ae04d8  a8b4c1808756  50aca0b19f7c  
osd.4                   ceph-node-1               running (26h)    94s ago   2d    60.3M    4096M  19.3.0-6008-g60ae04d8  a8b4c1808756  1560f80ba29a  
osd.5                   ceph-node-2               running (26h)    94s ago   2d    60.5M    1496M  19.3.0-6008-g60ae04d8  a8b4c1808756  6354be288c00  

```
**Tuned-profile example:**
```
[ceph: root@ceph-node-0 ~]# ceph orch tuned-profile ls
profile_name: 23-mon-tuned-profile
placement: ceph-node-0;ceph-node-1
settings:
  vm.swappiness: 10
  vm.user_reserve_kbytes: 116851
---

[ceph: root@ceph-node-0 ~]# ceph orch tuned-profile apply -i tuned-profile.yaml 
Error EAGAIN: WARNING: The orchestrator is paused. The command is deferred for execution until orchestrator is resumed. 
Saved tuned profile 23-mon-tuned-profile

[ceph: root@ceph-node-0 ~]# ceph orch tuned-profile ls
profile_name: 23-mon-tuned-profile
placement: ceph-node-0;ceph-node-1
settings:
  vm.swappiness: 10
---

```
**Alternate implementation:**
```
[ceph: root@ceph-node-0 ~]# ceph orch rm crash
Error EAGAIN: WARNING: The orchestrator is paused. Please resume the orchestrator and try again.

[ceph: root@ceph-node-0 ~]# ceph orch resume
[ceph: root@ceph-node-0 ~]# ceph orch ls
NAME                       PORTS  RUNNING  REFRESHED  AGE  PLACEMENT  
crash                                 4/4  4s ago     38m  *          
mgr                                   2/2  3s ago     2d   count:2    
mon                                   4/5  4s ago     2d   count:5    
osd.all-available-devices               7  4s ago     2d   *      
```

ceph-jenkins commented:
Can one of the admins verify this patch?

anthonyeleven commented on doc/cephadm/troubleshooting.rst:
```diff
@@ -36,6 +36,35 @@ These commands disable all ``ceph orch ...`` CLI commands. All
 they were before you ran these commands.
 
+In certain scenarios, commands may still execute even when the orchestrator is paused. 
+This behavior arises because these commands bypass the orchestrator's main serve loop and are not queued for asynchronous execution. 
```
s/serve/working/

Kushal-deb replied:
Adressed.

anthonyeleven commented on doc/cephadm/troubleshooting.rst:
```diff
@@ -36,6 +36,35 @@ These commands disable all ``ceph orch ...`` CLI commands. All
+This behavior arises because these commands bypass the orchestrator's main serve loop and are not queued for asynchronous execution. 
+Instead, they invoke low-level operations directly, allowing them to proceed regardless of the orchestrator's paused state.
+
+example: removing a Daemon
```
Example: removing a daemon

Kushal-deb replied:
Addressed

adk3798 commented on doc/cephadm/troubleshooting.rst:
```diff
@@ -36,6 +36,35 @@ These commands disable all ``ceph orch ...`` CLI commands. All
+   ceph orch daemon add
+   ceph orch host add <host>
+   ceph orch host rm <host>
+   ceph orch tuned-profile apply
```
if we're going to put `tuned-profile apply` in this list, I think we should include the normal `orch apply`. Afaik, they're handled similarly. In both cases the spec will get stored, but we won't act on the spec until the orchestrator is unpaused. Given how important the regular apply command is to cephadm operations, it's worth mentioning.

adk3798 commented on src/pybind/mgr/orchestrator/_interface.py:
```diff
@@ -159,6 +167,49 @@ def inner_cli_command(prefix: str) -> Callable[[FuncT], FuncT]:
+    return False
+
+
+def check_orchestrator_paused(status_method: str = '_status') -> Callable[[Callable[..., HandleCommandResult]], Callable[..., HandleCommandResult]]:
+    def decorator(func: Callable[..., HandleCommandResult]) -> Callable[..., HandleCommandResult]:
+        @wraps(func)
+        def wrapper(self: Any, *args: Any, **kwargs: Any) -> HandleCommandResult:
+            get_status_func = getattr(self, status_method, None)
+
+            if not callable(get_status_func):
+                raise RuntimeError(f"'{status_method}' method is not defined or callable on {self.__class__.__name__}")
+            status_result = get_status_func(self)
+
+            if not status_result.stdout.strip():
+                if func.__name__ == '_set_backend':
+                    return func(self, *args, **kwargs)
+                raise NoOrchestrator()
+
+            if "No orchestrator configured" in status_result.stdout:
+                if func.__name__ == '_set_backend':
+                    return func(self, *args, **kwargs)
+                raise NoOrchestrator()
+
+            if is_paused(status_result.stdout):
+                result = func(self, *args, **kwargs)
+                raise OrchestratorPaused(
+                    "WARNING: The orchestrator is paused.\n"
+                    "Some commands (e.g., 'ceph orch rm', 'ceph orch apply') will be deferred until the orchestrator is resumed, "
+                    "while others (e.g., 'ceph orch daemon rm', 'ceph orch daemon add') will execute immediately.\n"
+                    "Refer to the documentation for more details.\n\n"
+                    f"{result.stdout}"
+                )
+            return func(self, *args, **kwargs)
```
I don't feel like it's necessary to have this whole setup with result parsing for calling a function defined in another file in the same directory. If the reason this needs to be done is that `_status` has that `_cli_read_command` header on it and therefore can't be invoked directly, we could make the current `_status` into something of a stub function and move it's actual contents to another function that can be called from both there and here. That's all assuming we can't just call the `available` function the way `_status` already does here and then just do something like that function is already doing e.g.
```
avail, why, module_details = self.available()
result: Dict[str, Any] = {
    "available": avail,
    "backend": o,
}

if avail:
     result.update(module_details)
else:
     <indicate we don't know>
if 'paused' in result and result['paused']:
    <indicate is paused>
elif 'paused' in result:
    <indicate not paused>
else:
    <indicate we don't know>
```
worth noting in that example, both the "don't know" cases would also be "don't know" cases when invoking `_status` as we are here, since they use the same underlying mechanism.
